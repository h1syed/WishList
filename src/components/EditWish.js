import React, {Component} from "react";
import { Auth, API, graphqlOperation } from 'aws-amplify';
import { updateWish } from "../graphql/mutations";

class EditWish extends Component {

    state = {
        show: false,
        id: "",
        wishOwnerId: "",
        wishOwnerUsername: "",
        wishTitle: "",
        wishBody: "",
        wishData: {
             wishTitle: this.props.wishTitle,
             wishBody: this.props.wishBody
        }
    }

    handleModal = () => {
        this.setState({ show: !this.state.show});
        document.body.scrollTop = 0;
        document.documentElement.scrollTop = 0;
   }

   handleUpdateWish = async (event) => {
    event.preventDefault();
    const {username} = await Auth.currentUserInfo();
    const input = {
         id: this.props.id,
         wishOwnerId: this.state.wishOwnerId,
         wishOwnerUsername: this.state.wishOwnerUsername,
         wishTitle: this.state.wishData.wishTitle,
         wishBody: this.state.wishData.wishBody

    }

    await API.graphql(graphqlOperation(updateWish, { input },{ owner: username }));

    //force close the modal 
    this.setState({ show: !this.state.show});

}

handleTitle = event => {
    this.setState({
        wishData: {...this.state.wishData, wishTitle: event.target.value}
         
    });
}

handleBody = event => {
    this.setState({ wishData: {...this.state.wishData,
     wishBody: event.target.value}})
}

    componentDidMount = async () => {

        await Auth.currentUserInfo()
            .then(user => {
                 this.setState({
                     wishOwnerId: user.attributes.sub,
                     wishOwnerUsername: user.username 
                 })
            })
         
    }

    render() {
        return (
            <>
                { this.state.show && (
                 <div className="modal">
                      <button className="close"
                         onClick={this.handleModal}>
                          X
                      </button>

                      <form className="add-wish"
                         onSubmit={(event) => this.handleUpdateWish(event)}>

                             <input style={{fontSize: "19px"}}
                                  type="text" placeholder="Title"
                                  name="wishTitle"
                                  value={this.state.wishData.wishTitle}
                                  onChange={this.handleTitle} />

                             <input 
                                style={{height: "150px", fontSize: "19px"}}
                                type="text"
                                name="wishBody"
                                value={this.state.wishData.wishBody}
                                onChange={this.handleBody}
                                />

                             <button className="wish-button">Update Wish</button>


                      </form>

                      
                 </div>
             )
             }


                    
                <button className="wish-button" onClick={this.handleModal}>Edit</button>
            </>
        )
    }
}
export default EditWish;