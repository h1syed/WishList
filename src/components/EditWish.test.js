import React, { Component } from "react";
import { shallow } from "enzyme";
import { API, Auth } from "aws-amplify";
import EditWish from "./EditWish";

describe("Edit wish component", () => {
    it("renders properly", () => {
        const wrapper = shallow(<EditWish/>);
        expect(wrapper).toMatchSnapshot();
    });

    it("componentDidMount" , async () => {
        const wish = {
            wishTitle: "",
            wishBody: ""
        }
        const wrapper = shallow(<EditWish {...wish}/>);
        const instance = wrapper.instance();
        const user = {
            attributes: {
                sub: "123"
            },
            username: "Bob"
        }
        const spy = jest.spyOn(Auth, "currentUserInfo").mockReturnValue(Promise.resolve(user));
        await instance.componentDidMount();
        expect(instance.state).toEqual({
            show: false,
            id: "",
            wishOwnerId: "123",
            wishOwnerUsername: "Bob",
            wishTitle: "",
            wishBody: "",
            wishData: {
                wishTitle: "",
                wishBody: ""
           }
        });
        spy.mockRestore();
    });

describe("modal", () => {
    it("close button calls handleModal onClick", () => {
        const wrapper = shallow(<EditWish/>).setState({show: true});
        const spy = jest.spyOn(wrapper.instance(), "handleModal");
        wrapper.instance().forceUpdate()
        wrapper.find({ className: "close"}).simulate("click");
        expect(spy).toHaveBeenCalledTimes(1);
        expect(wrapper.instance().state.show).toBe(false);
        spy.mockRestore();
    });
});

describe("form", () => {


    it("calls handleUpdateWish onSubmit when change is submitted", async () => {
        const wrapper = shallow(<EditWish/>).setState({show: true});
        const mockPreventDefault = jest.fn();
        const spy = jest.spyOn(wrapper.instance(), "handleUpdateWish");
        const spy1 = jest.spyOn(Auth, "currentUserInfo").mockReturnValue(Promise.resolve(""))
        const spy2 = jest.spyOn(API, "graphql").mockReturnValue(Promise.resolve());
        wrapper.instance().forceUpdate();
        const mockEvent = {
            preventDefault: mockPreventDefault
        }
        wrapper.find({ className: "add-wish"}).simulate("submit", mockEvent);
        expect(spy).toHaveBeenCalledWith(mockEvent);
        spy.mockRestore();
        spy1.mockRestore();
        wrapper.setState({ show: false});
    });

    it("calls handleTitle onChange of input for wishTitle", () => {    
        const wrapper = shallow(<EditWish/>).setState({ show: true });
        const spy = jest.spyOn(wrapper.instance(), "handleTitle");
        wrapper.instance().forceUpdate();
       const mockEvent= {
            target: {
                name: "wishTitle",
                value: "This is my new wish"
            }
        }
        wrapper.find({ name: "wishTitle"}).simulate("change",mockEvent);
        expect(spy).toHaveBeenCalledWith(mockEvent);
        wrapper.setState({ show: false });
        spy.mockRestore();
    });

    it("calls handleBody onChange of input for wishBody", () => {
        const wrapper = shallow(<EditWish/>).setState({ show: true });
        const spy = jest.spyOn(wrapper.instance(), "handleBody");
        wrapper.instance().forceUpdate();
        const mockEvent = {
            target: {
                value: "This is an event"
            }
        }
        wrapper.find({ name: "wishBody"}).simulate("change", mockEvent);
        expect(spy).toHaveBeenCalledWith(mockEvent);
        wrapper.setState({ show: false });
        spy.mockRestore();
    });
});
});