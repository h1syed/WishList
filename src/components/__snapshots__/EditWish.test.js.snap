// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit wish component renders properly 1`] = `ShallowWrapper {}`;
